### ISSUE TITLE
Use only a config file to load the parameters

### ISSUE THREAD

**Issue by davidducos** [maxbube/mydumper#318]
It was requested by @druud in https://github.com/maxbube/mydumper/issues/84

So, as we are adding more parameters which defaults values are not ideal for every use case, will be simpler for users to just keep the data in a configuration file.

**Comment by davidducos** [comment]
Implemented in https://github.com/maxbube/mydumper/tree/config_file

**Comment by davidducos** [comment]
We can use different sections, not just: `[mydumper]` or `[myloader]` as the initial intention, we could also [implement the SET SESSION variable](https://github.com/maxbube/mydumper/issues/50) in a new section like `[mydumper_session_variables]` or `[myloader_session_variables]`.

**Comment by druud** [comment]
The initial #84 also mentioned '[rebuild]' as a group name.

One of the goals is to be able to use mydumper/myloader to rebuild a table as fast as possible, 
so to be able to make a dump that is optimized for the **target** structure of a table.
For that, each target .ibd file (and a partitioned table has several) must have a single thread writing to it, in perfect PK-order.

**Comment by davidducos** [comment]
Hi @druud, the [rebuild] group that you proposed has been renamed to [mydumper], we added also [myloader] to allow users to maintain a single file.

About the pk insert order, I would like to make a better analysis as I think that latest implementation will be better as it orders the files. Doing single threaded inserts goes against the idea of this project, which is to perform multi threaded inserts even per table to speed up the inserts.

**Comment by druud** [comment]
Re. "multi threaded inserts even per table to speed up the inserts", that is a common misconception.

First, it is not "per table", but "per ibd-file". Second, a single thread filling up an ibd-file, in perfect PK-order, and without secondary indexes, is in many aspects the most optimal way to fill up a table.

For the highest speed, use something close to the default --tab dump format of mysqldump, for example:
`mysql -BNe 'SET @@time_zone="+0:0"; SELECT * FROM MyTable FORCE INDEX (PRIMARY) LIMIT 10'`
Be aware that assumes that the columns structure remains unchanged.
Also be aware that it escapes embedded \t, \n, \0 and \\.

Inserting in perfect appending order (using LOAD DATA INFILE) is the lightest, which makes it feasible to fill many ibd-files in parallel, without risking IO- and memory-usage-overload.
Be aware that each table partition (of a partitioned table) has its own ibd-file.

Re-apply all secondary indexes only after filling up the tables, which can (if really needed) even be done 'inplace', so with the table already in use.

**Comment by davidducos** [comment]
Hi @druud ,

myloader is not aware about the ibd files, just tables when we talked about data. You are totally right that a single thread is the most optimal way, but it is not the fastest. Take into account that single threaded inserts per table on multiple tables has lower insert ratio per table than multithreaded inserts in a single table. The most important use cases for MyDumper is when you have a couple of tables with million or billion rows that you want to restore as fast as possible, and the fastest way of doing it, is to fill up the largest tables as fast as possible to start the index creation. But again, I will review how bad is myloader doing it. 

I did some test with LOAD DATA and I found that the performance was very similar to multithreaded inserts, I'm still working in that piece of code.
There is another ticket to work with partitioned tables, but honestly, I didn't have much time push it. 

I don't understand your latest comment, myloader is applying all the secondary indexes at once when it is possible (it is not possible when you have more than one FULLTEXT indexes), and the alters starts after all inserts had finished, it is using fast index creation, so, they are inplace.

**Comment by druud** [comment]
Re. "myloader is not aware about the ibd files, just tables when we talked about data.", 
it is easy to find out: there is the @@innodb_file_per_table setting, 
and there is the desired target table layout, which can be partitioned, which then means multiple ibd-files for that table.
So basically going from tables to their ibd-files (or call it "independent write targets") is an easy step.

Re. "Take into account that single threaded inserts per table on multiple tables has lower insert ratio per table than multithreaded inserts in a single table.", you keep mentioning that, but let's first come up with some real benchmarks before  we can agree to believe in that. And then also have the case with many tables being filled concurrently.

Be aware of the overhead involved in writing rows in non-PK-order, like page splits etc. Assuming 2 threads only, that both progress in PK-order, then the top-page of thread-1 and the bottom-page of thread-2 can have very unhealthy interactions, that are tricky to prevent.

**Comment by davidducos** [comment]
Hi @druud,

Yes, checking the @@innodb_file_per_table is the simplest use case, as we might have tablespaces. For sure, this will need more coding.

I did some benchmarks more than a month ago, I will be adding a page in the wiki, later this afternoon with some testings that I did about the primary plus a PR that I will be publishing later today. I can bet that you will love it! 

Cheers! and thanks for all your feedback!

**Comment by druud** [comment]
To add some color: I only use partition formulas that can easily be expressed in a WHERE-clause.
For example `PARTITIONED BY HASH(id) PARTITIONS 8` means that `id MOD 8` is the partition number.

Same for UUIDs, partitioning on selected bits from one of the bytes in the value, to spread the values over N partitions.
I generally prefer powers of 2 for the number of partitions.

This all to make it easy for the client-application to write rows directly into the dedicated partition, to gain performance.

Another special example:
`PARTITION BY HASH ((id DIV 4194304)) PARTITIONS 64`.
This to have a chunk of new rows inside the same partition, so in a limited set of memory pages.
